<!DOCTYPE HTML>
<html>
<head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8">
    <title>Exception</title>
    <link rel="stylesheet" href="/style.css">
    <script src="/script.js" defer></script>
</head>
<body>
<div><a href="../Core.html" class="backButton"></a></div>
<div class="bodyStyle">
    <h1>Exception</h1>

    <div class="todo">
        <h2>Проработать:</h2>
        <ol>
            <li><a href="https://javarush.ru/groups/posts/isklyucheniya-java">Исключения в Java</a></li>
            <li><a href="https://habr.com/ru/company/golovachcourses/blog/223821/">Исключения в Java, Часть I (try-catch-finally)</a></li>
            <li><a href="https://habr.com/ru/company/golovachcourses/blog/225585/">https://habr.com/ru/company/golovachcourses/blog/225585/#1</a></li>
            <li><a href="https://habr.com/ru/post/540102/">Вы всё ещё ловите исключения? Тогда мы к вам</a></li>
        </ol>
    </div>

    <h2>Два главных типа:</h2>
    <ol>
        <li>checked - компилятор их проверяет.</li>
        <li>unchecked - компилятор их не проверяет.</li>
    </ol>

    <h2>Основные определения:</h2>
    <ol>
        <li>Exception(Исключение) - это нештатная ситуация, ошибка во время выполнения программы.</li>
        <li>Error - не обрабатываемые исключения тк после из возникновений нет гарантий что программа будет
            выполняться корректно.
        </li>
    </ol>

    <h2>Иерархия исключений:</h2>
    <img src="ExceptionHierarchy.png" width="1100" height="500">

    <h2>Способы обработки исключений</h2>
    <ol>
        <li>throws - ключевое слово которое, используется при определении метода. Указывает что метод может выкинуть
            ошибку.
        </li>
        <li>try - catch - finally
            <pre>
try(try-with-resources){}
catch(Exception e) {}
catch(ClassCastException | ArrayStoreException e1) {}
finally {}
            </pre>
            <p>try-with-resources - дает возможность объявлять один или несколько ресурсов в блоке try, которые
                будут закрыты автоматически без использования finally блока. Проверяет наличие интерфейса AutoCloseable
                или Closeable. И выполняет метод close.
                <pre>
try (BufferedReader br = new BufferedReader(new FileReader(path))) {
    return br.readLine();
}
            </pre>
            </p>
        </li>
    </ol>

    <h2>Иерархия исключений:</h2>
    <ul>
        <li>ArithmeticException - арифметическая ошибка: деление на ноль и тд.</li>
        <li>ArrayIndexOutOfBoundsException - индекс массива находится вне границ.</li>
        <li>ArrayStoreException - назначение элементу массива несовместимого типа.</li>
        <li>ClassCastException - недопустимое приведение типов.</li>
        <li>ConcurrentModificationException - некорректная модификация коллекции.</li>
        <li>IllegalArgumentException - при вызове метода использован незаконный аргумент.</li>
        <li>IllegalMonitorStateException - незаконная операция монитора на разблокированном экземпляре.</li>
        <li>IllegalStateException - среда или приложение находятся в некорректном состоянии.</li>
        <li>IllegalThreadStateException - требуемая операция не совместима с текущим состоянием потока.</li>
        <li>IndexOutOfBoundsException - некоторый тип индекса находится вне границ.</li>
        <li>NegativeArraySizeException - массив создавался с отрицательным размером.</li>
        <li>NullPointerException - недопустимое использование нулевой ссылки.</li>
        <li>NumberFormatException - недопустимое преобразование строки в числовой формат.</li>
        <li>StringIndexOutOfBoundsException - попытка индексации вне границ строки.</li>
        <li>UnsupportedOperationException - встретилась неподдерживаемая операция.</li>
    </ul>

    <h2>Вопросы на собеседование:</h2>
    <ol>
        <li>Если наследовать класс с throws. Наследник должен отлавливать throws не выше по иерархии чем у родителя.</li>
        <li>Если используются несколько catch нужно писать их отлавливаемые исключения в порядке возвышения по иерархии.</li>
        <li>Можно написать try - finally без блока catch.</li>
        <li>try-with-resources - объект должен реализовывать Interface Closeable.</li>
        <li>Условия не выполнения finally:
            <ol>
                <li>System.exit(int)</li>
                <li>Сбой в JVM</li>
                <li>Exception в finally</li>
                <li>Death of Daemon thread</li>
            </ol>
        </li>
    </ol>
</div>
</body>
<script>setBackButtonHeight();</script>
</html>